File: app/src/main/java/com/renpytool/UpdateChecker.java
package com.renpytool;

import android.util.Log;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;

/**
 * Checks for updates from GitHub releases
 */
public class UpdateChecker {

    private static final String TAG = "UpdateChecker";
    private static final String GITHUB_API_URL = "https://api.github.com/repos/Sirpixelalot/Rentool/releases/latest";
    private static final int TIMEOUT_MS = 10000;  // 10 seconds

    public interface UpdateCheckCallback {
        void onUpdateAvailable(VersionInfo versionInfo);
        void onNoUpdateAvailable();
        void onCheckFailed(String error);
    }

    /**
     * Check for updates on GitHub releases
     * This method should be called from a background thread
     */
    public static void checkForUpdates(String currentVersion, UpdateCheckCallback callback) {
        Log.d(TAG, "Checking for updates... Current version: " + currentVersion);
        HttpURLConnection connection = null;
        BufferedReader reader = null;

        try {
            // Create connection
            URL url = new URL(GITHUB_API_URL);
            Log.d(TAG, "Connecting to: " + GITHUB_API_URL);
            connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");
            connection.setConnectTimeout(TIMEOUT_MS);
            connection.setReadTimeout(TIMEOUT_MS);
            connection.setRequestProperty("Accept", "application/vnd.github.v3+json");

            // Check response code
            int responseCode = connection.getResponseCode();
            Log.d(TAG, "Response code: " + responseCode);
            if (responseCode != HttpURLConnection.HTTP_OK) {
                Log.e(TAG, "HTTP error: " + responseCode);
                callback.onCheckFailed("HTTP error: " + responseCode);
                return;
            }

            // Read response
            reader = new BufferedReader(new InputStreamReader(connection.getInputStream()));
            StringBuilder response = new StringBuilder();
            String line;
            while ((line = reader.readLine()) != null) {
                response.append(line);
            }

            // Parse JSON response
            JsonObject jsonResponse = JsonParser.parseString(response.toString()).getAsJsonObject();

            // Extract version information
            String tagName = jsonResponse.has("tag_name") ? jsonResponse.get("tag_name").getAsString() : null;
            String body = jsonResponse.has("body") ? jsonResponse.get("body").getAsString() : "";
            String publishedAt = jsonResponse.has("published_at") ? jsonResponse.get("published_at").getAsString() : "";

            // Get download URL from assets (if available)
            String downloadUrl = jsonResponse.has("html_url") ? jsonResponse.get("html_url").getAsString() : "";

            Log.d(TAG, "Latest version from GitHub: " + tagName);
            Log.d(TAG, "Download URL: " + downloadUrl);

            // Check if tag name is valid
            if (tagName == null || tagName.isEmpty()) {
                Log.e(TAG, "Invalid tag name from GitHub");
                callback.onCheckFailed("Invalid response from GitHub");
                return;
            }

            // Create version info object
            VersionInfo versionInfo = new VersionInfo(tagName, downloadUrl, body, publishedAt);

            // Compare versions
            Log.d(TAG, "Comparing: " + versionInfo.getVersionNumber() + " vs " + currentVersion);
            if (versionInfo.isNewerThan(currentVersion)) {
                Log.d(TAG, "Update available!");
                callback.onUpdateAvailable(versionInfo);
            } else {
                Log.d(TAG, "No update available");
                callback.onNoUpdateAvailable();
            }

        } catch (Exception e) {
            // Silent fail - don't bother user with network errors
            Log.e(TAG, "Update check failed", e);
            callback.onCheckFailed("Network error: " + e.getMessage());

        } finally {
            // Clean up resources
            try {
                if (reader != null) reader.close();
            } catch (Exception e) {
                // Ignore
            }
            if (connection != null) {
                connection.disconnect();
            }
        }
    }
}

File: app/src/main/java/com/renpytool/UpdateChecker.kt
package com.renpytool

import android.util.Log
import com.google.gson.JsonParser
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL

/**
 * Checks for updates from GitHub releases using Kotlin coroutines
 */
object UpdateChecker {

    private const val TAG = "UpdateChecker"
    private const val GITHUB_API_URL = "https://api.github.com/repos/Sirpixelalot/Rentool/releases/latest"
    private const val TIMEOUT_MS = 10000  // 10 seconds

    sealed class UpdateResult {
        data class UpdateAvailable(val versionInfo: VersionInfo) : UpdateResult()
        object NoUpdateAvailable : UpdateResult()
        data class CheckFailed(val error: String) : UpdateResult()
    }

    /**
     * Check for updates on GitHub releases
     * This is a suspend function that runs on IO dispatcher
     */
    suspend fun checkForUpdates(currentVersion: String): UpdateResult = withContext(Dispatchers.IO) {
        Log.d(TAG, "Checking for updates... Current version: $currentVersion")
        var connection: HttpURLConnection? = null
        var reader: BufferedReader? = null

        try {
            // Create connection
            val url = URL(GITHUB_API_URL)
            Log.d(TAG, "Connecting to: $GITHUB_API_URL")
            connection = (url.openConnection() as HttpURLConnection).apply {
                requestMethod = "GET"
                connectTimeout = TIMEOUT_MS
                readTimeout = TIMEOUT_MS
                setRequestProperty("Accept", "application/vnd.github.v3+json")
            }

            // Check response code
            val responseCode = connection.responseCode
            Log.d(TAG, "Response code: $responseCode")
            if (responseCode != HttpURLConnection.HTTP_OK) {
                Log.e(TAG, "HTTP error: $responseCode")
                return@withContext UpdateResult.CheckFailed("HTTP error: $responseCode")
            }

            // Read response
            reader = BufferedReader(InputStreamReader(connection.inputStream))
            val response = reader.readText()

            // Parse JSON response
            val jsonResponse = JsonParser.parseString(response).asJsonObject

            // Extract version information
            val tagName = jsonResponse.get("tag_name")?.asString
            val body = jsonResponse.get("body")?.asString ?: ""
            val publishedAt = jsonResponse.get("published_at")?.asString ?: ""
            val downloadUrl = jsonResponse.get("html_url")?.asString ?: ""

            Log.d(TAG, "Latest version from GitHub: $tagName")
            Log.d(TAG, "Download URL: $downloadUrl")

            // Check if tag name is valid
            if (tagName.isNullOrEmpty()) {
                Log.e(TAG, "Invalid tag name from GitHub")
                return@withContext UpdateResult.CheckFailed("Invalid response from GitHub")
            }

            // Create version info object
            val versionInfo = VersionInfo(tagName, downloadUrl, body, publishedAt)

            // Compare versions
            Log.d(TAG, "Comparing: ${versionInfo.versionNumber} vs $currentVersion")
            if (versionInfo.isNewerThan(currentVersion)) {
                Log.d(TAG, "Update available!")
                UpdateResult.UpdateAvailable(versionInfo)
            } else {
                Log.d(TAG, "No update available")
                UpdateResult.NoUpdateAvailable
            }

        } catch (e: Exception) {
            // Silent fail - don't bother user with network errors
            Log.e(TAG, "Update check failed", e)
            UpdateResult.CheckFailed("Network error: ${e.message}")

        } finally {
            // Clean up resources
            reader?.close()
            connection?.disconnect()
        }
    }

    /**
     * Legacy callback interface for backward compatibility
     */
    interface UpdateCheckCallback {
        fun onUpdateAvailable(versionInfo: VersionInfo)
        fun onNoUpdateAvailable()
        fun onCheckFailed(error: String)
    }

    /**
     * Legacy method with callback for backward compatibility
     */
    @Suppress("unused")
    suspend fun checkForUpdates(currentVersion: String, callback: UpdateCheckCallback) {
        when (val result = checkForUpdates(currentVersion)) {
            is UpdateResult.UpdateAvailable -> callback.onUpdateAvailable(result.versionInfo)
            is UpdateResult.NoUpdateAvailable -> callback.onNoUpdateAvailable()
            is UpdateResult.CheckFailed -> callback.onCheckFailed(result.error)
        }
    }
}
